sql: support for RETURNING *

RaduBerinde opened the pull request:
This change adds support for RETURNING \* in inserts and deletes. For updates we
currently only expand the updated columns; this is related to an existing
problem (we can't return other columns even explicitly; filed #4645).

Fixes #4593.

@knz @mjibson

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4647)

<!-- Reviewable:end -->

tamird commented:
Haven't reviewed the implementation yet, will do that this afternoon. Glad to see this fixed, though!

---

Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/logic_test.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMKMmEUrJ4819CkJo:-KBOMKMnUxx_oj4dlGVz:-378761441) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/logic_test.go#L555)):*
why do you need to do this? seems odd to split column names on whitespace.

---

*[sql/testdata/insert, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMUuLb_yKcUz3A715:-KBOMUuLb_yKcUz3A716:2119742349) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/insert#L362)):*
is this different from the `INSERT INTO return VALUES (1) RETURNING *` case?

---

*[sql/testdata/update, line 190 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMgm5TuTk8-dGV2Pz:-KBOMgm5TuTk8-dGV2Q-:-1143692787) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/update#L190)):*
nit: multiple spaces here

---

*[sql/testdata/update, line 197 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMiP0tOz3ET2Lb6k4:-KBOMiP0tOz3ET2Lb6k5:1558766889) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/update#L197)):*
here too

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOMqs7nqtTwkeQu1-A:-1767963460)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/logic_test.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMKMmEUrJ4819CkJo:-KBOOApmTrx9ekDVbiw1:-1853079677) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/logic_test.go#L555)):*
To support whitespaces in column names, it's the same odd way we support whitespaces in results (copied from below).  Without this we get an error like
```
	logic_test.go:654: testdata/insert:355: expected:
		"a"  "b"  "a"  
		"+"  "b"  "1"  
		"2"  "3"  "3"  
		"4"  "7"  
		but found:
		"a"  "b"  "a + b"  
		"1"  "2"  "3"      
		"3"  "4"  "7"      
```

---

*[sql/testdata/insert, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMUuLb_yKcUz3A715:-KBOOo8fO-mzW75DMgR5:252178691) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/insert#L362)):*
In that one we are only inserting to one column. I added the `VALUES (1)` one to check if this is a problem (like it is for updates).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647)*
<!-- Sent from Reviewable.io -->

knz commented:
LGTM

---

Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOZiKlKfevsMvOdBwR:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/logic_test.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMKMmEUrJ4819CkJo:-KBOYRratvbQAonnV7p3:921324127) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/logic_test.go#L555)):*
OK. Can you update the comment to refer to whitespace in column names rather than results?

---

*[sql/returning.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOoeEOfGZN3ggTAwzu:-KBOoeEPkORcnNLSJ0XZ:176983786) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/returning.go#L38)):*
super nit: we usually name functions that return values (rather than pointers) `makeFoo` rather than `newFoo`

---

*[sql/returning.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOooHNiaCuM9RJ2vVi:-KBOooHNiaCuM9RJ2vVj:-1623933175) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/returning.go#L53)):*
remove this newline

---

*[sql/select.go, line 373 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOYeBF2Jl9lHkety4T:-KBOYeBF2Jl9lHkety4U:-1613931549) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L373)):*
s/prefix/suffix/

nit: golint doesn't care because this isn't exported, but for consistency please start the comment with the function name.

---

*[sql/select.go, line 378 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZPGVBoboY2jrDEO5:-KBOZPGVBoboY2jrDEO6:683365926) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L378)):*
the named returns here serve as nice documentation, but it makes me nervous about naked `return`s. doesn't look like you're using any, but I just wanted to point out the possibility for abuse.

---

*[sql/select.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZKmGqB9kAbMB6MeS:-KBOZKmGqB9kAbMB6MeT:-667511596) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L379)):*
nit: remove this newline

---

*[sql/select.go, line 399 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZhR9jLS9UO04_lJB:-KBOZhR9jLS9UO04_lJC:-243197048) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L399)):*
nit: doesn't look like this is used beyond the next `if` block, could be nice to reduce its lifetime:
```
if tablename := ...; tablename != ""....
```

---

*[sql/testdata/insert, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMUuLb_yKcUz3A715:-KBOYM39_C95KRL4PLCe:-840806217) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/insert#L362)):*
OK. Maybe put these cases next to each other, since it's not clear why they're ordered as they are now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOp-PPoQ64OwM-0XSx:676834233)*
<!-- Sent from Reviewable.io -->

knz commented:
Wait please add a test using Prepare as well. The lesson learned with #4610 is that prepare != execute.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOqHO8Ia5EupuAZt9X:-1408315192)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I can't until #4635 is merged

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/select.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZKmGqB9kAbMB6MeS:-KBOqKWz2nnoioSjFJ_H:345274781) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L379)):*
I'd like to keep it, it separates the function declaration from the code. I wish gofmt would append two tabs instead of one when we break up a declaration line.

---

*[sql/testdata/insert, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOMUuLb_yKcUz3A715:-KBOqu8K1v3PuLZ2TAGj:-1402129143) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/testdata/insert#L362)):*
Moved and added a comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOtm4rYtBZy_n4-4OF:385793832)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: support for RETURNING *

This change adds support for `RETURNING *` in inserts and deletes. For updates
we currently only expand the updated columns; this is related to an existing
problem (we can't return other columns even explicitly, #4368).

Fixes #4593.

mjibson commented:
LGTM

---

Review status: 2 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/select.go, line 382 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOv6OnHdWOo8GUHPSq:-KBOv6OoJsgauRKf37YU:-1252239536) ([raw file](https://github.com/cockroachdb/cockroach/blob/db75008c09af697681874c868abd5a847f9e4404/sql/select.go#L382)):*
Any problem with using a naked return here? Some people dislike them, which is a fine argument.

---

*[sql/select.go, line 384 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOvDgpjRl_dRbKqBmc:-KBOvDgqoLPytz2diHXI:433541382) ([raw file](https://github.com/cockroachdb/cockroach/blob/db75008c09af697681874c868abd5a847f9e4404/sql/select.go#L384)):*
Could omit the `:` here because err is already declared and in a parent scope, then use a naked return below. Or not, I don't feel strongly.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-:-KBOxJyhL3M81XQViy18:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/select.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZKmGqB9kAbMB6MeS:-KBOuHmll4ZaIRRYzicY:-1213286753) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L379)):*
I appreciate that, but I would prefer to stick to the pervasive style for now. We should discuss this in the context of the upcoming style guide. cc @bdarnell 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/select.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZKmGqB9kAbMB6MeS:-KBOy1R9qHXxwCWPYN0D:952114754) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L379)):*
The situation of a multi-line function definition is anything but pervasive, but ok. I think we're wasting too much time with these nits btw.

---

*[sql/select.go, line 382 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOv6OnHdWOo8GUHPSq:-KBOxhwu_3B3oqv4B_ec:-1503985459) ([raw file](https://github.com/cockroachdb/cockroach/blob/db75008c09af697681874c868abd5a847f9e4404/sql/select.go#L382)):*
I think they can be error prone (I put variable names in just for documentation). The official go tour says they should be only used in short functions.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 8 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/select.go, line 379 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647#-KBOZKmGqB9kAbMB6MeS:-KBPuCW0-7lwnZr4j2oV:-316358726) ([raw file](https://github.com/cockroachdb/cockroach/blob/559f75793d89a86724b1bc9ec7004f591d02e2c2/sql/select.go#L379)):*
Noted. FWIW, I agree with Radu: a blank line can add clarity when the declaration spans multiple lines (unless gofmt were changed to double-indent continuation lines). 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4647)*
<!-- Sent from Reviewable.io -->
